src/main/java/org/littleshoot/proxy/impl/ClientToProxyConnection.java

```diff
@@ -516,17 +516,19 @@ protected void connected() {
         recordClientConnected();
     }
 
-    @Override
-    protected void timedOut() {
-        boolean clientReadMoreRecentlyThanServer =
-                currentServerConnection == null
-                        || this.lastReadTime > currentServerConnection.lastReadTime;
-        if (clientReadMoreRecentlyThanServer) {
-            LOG.debug("Server timed out: {}", currentServerConnection);
+    void timedOut(ProxyToServerConnection serverConnection) {
+        if (currentServerConnection == serverConnection && this.lastReadTime > currentServerConnection.lastReadTime) {
+            // the idle timeout fired on the active server connection. send a timeout response to the client.
+            LOG.warn("Server timed out: {}", currentServerConnection);
             currentFilters.serverToProxyResponseTimedOut();
             writeGatewayTimeout(currentRequest);
```

Comment:
Splitting this into two methods allows control over how many times writeGatewayTimeout gets called during proxy2server timeout.  If you make this change https://github.com/adamfisk/LittleProxy/pull/350/commits/55fd04d744e3db54b5a37206f682affb5465f56a the KeepAlive tests will pass and it shows that currently it is called twice, which can confuse some HTTP clients very much.